File: modules/expo-native-mqtt/android/src/main/java/expo/modules/nativemqtt/ExpoNativeMqttModule.kt
package expo.modules.nativemqtt

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import com.hivemq.client.mqtt.mqtt3.Mqtt3AsyncClient
import com.hivemq.client.mqtt.mqtt3.Mqtt3Client
import com.hivemq.client.mqtt.datatypes.MqttQos
import com.hivemq.client.mqtt.MqttGlobalPublishFilter
import com.hivemq.client.mqtt.mqtt3.message.connect.connack.Mqtt3ConnAck
import android.os.Handler
import android.os.Looper
import java.util.UUID

class ExpoNativeMqttModule : Module() {
    private var client: Mqtt3AsyncClient? = null
    private var storedOptions: Map<String, Any> = emptyMap()
    private var subscribedTopics = mutableMapOf<String, Int>() // topic -> qos
    private var isDisconnectingManually = false
    private val handler = Handler(Looper.getMainLooper())
    private var reconnectRunnable: Runnable? = null
    private var storedUsername: String? = null
    private var storedPassword: String? = null

    override fun definition() = ModuleDefinition {
        Name("ExpoNativeMqtt")

        Events(
            "onMqttConnected",
            "onMqttDisconnected",
            "onMqttError",
            "onMqttMessageReceived",
            "onMqttSubscribed",
            "onMqttUnsubscribed",
            "onMqttReconnecting"
        )

        AsyncFunction("connect") { brokerUrl: String, username: String?, password: String?, options: Map<String, Any>, promise: Promise ->
            try {
                // Parse options
                storedOptions = options
                val clientId = options["clientId"] as? String ?: "expo-mqtt-${UUID.randomUUID()}"
                val cleanSession = options["cleanSession"] as? Boolean ?: false
                val keepAlive = (options["keepAlive"] as? Number)?.toInt() ?: 60

                // Parse Broker URL
                // Format expects tcp://host:port or ssl://host:port
                val cleanUrl = brokerUrl.replace("tcp://", "").replace("ssl://", "")
                val parts = cleanUrl.split(":")
                val host = parts.getOrNull(0) ?: brokerUrl
                val port = parts.getOrNull(1)?.toInt() ?: 1883

                val builder = Mqtt3Client.builder()
                    .identifier(clientId)
                    .serverHost(host)
                    .serverPort(port)
                    .addDisconnectedListener { context ->
                        val error = context.cause
                        sendEvent("onMqttDisconnected", mapOf("error" to (error?.message ?: "Disconnected")))
                        if (!isDisconnectingManually) {
                            scheduleReconnect()
                        }
                    }

                // SSL Setup
                if (brokerUrl.startsWith("ssl://")) {
                    builder.sslWithDefaultConfig()
                }

                client = builder.buildAsync()

                // Global Publish listener
                client?.publishes(MqttGlobalPublishFilter.ALL) { publish ->
                    val payload = String(publish.payloadAsBytes)
                    sendEvent("onMqttMessageReceived", mapOf(
                        "topic" to publish.topic.toString(),
                        "payload" to payload,
                        "qos" to publish.qos.code,
                        "retained" to publish.isRetain
                    ))
                }

                // Connect builder
                storedUsername = username
                storedPassword = password

                val connectBuilder = client?.connectWith()
                    ?.cleanSession(cleanSession)
                    ?.keepAlive(keepAlive)

                if (!storedUsername.isNullOrEmpty() && !storedPassword.isNullOrEmpty()) {
                    connectBuilder?.simpleAuth()
                        ?.username(storedUsername!!)
                        ?.password(storedPassword!!.toByteArray())
                        ?.applySimpleAuth()
                }

                isDisconnectingManually = false

                connectBuilder?.send()?.whenComplete { ack: Mqtt3ConnAck?, throwable: Throwable? ->
                    if (throwable != null) {
                        promise.reject("CONNECT_FAILED", throwable.message, throwable)
                        sendEvent("onMqttError", mapOf("error" to throwable.message))
                    } else {
                        promise.resolve("Connected")
                        sendEvent("onMqttConnected", mapOf("status" to "connected"))
                        resubscribeAll()
                    }
                }
            } catch (e: Exception) {
                promise.reject("CONNECT_ERROR", e.message, e)
            }
        }

        AsyncFunction("disconnect") { promise: Promise ->
            isDisconnectingManually = true
            reconnectRunnable?.let { handler.removeCallbacks(it) }
            reconnectRunnable = null
            
            client?.disconnect()?.whenComplete { _, _ ->
                promise.resolve("Disconnected")
            } ?: promise.resolve("No active client")
        }

        AsyncFunction("subscribe") { topic: String, qos: Int, promise: Promise ->
            subscribedTopics[topic] = qos
            val mqttQos = MqttQos.fromCode(qos) ?: MqttQos.AT_MOST_ONCE
            
            client?.subscribeWith()
                ?.topicFilter(topic)
                ?.qos(mqttQos)
                ?.send()
                ?.whenComplete { _, throwable ->
                    if (throwable != null) {
                        promise.reject("SUBSCRIBE_FAILED", throwable.message, throwable)
                    } else {
                        promise.resolve("Subscribed")
                        sendEvent("onMqttSubscribed", mapOf("topic" to topic))
                    }
                }
        }

        AsyncFunction("unsubscribe") { topic: String, promise: Promise ->
            subscribedTopics.remove(topic)
            
            client?.unsubscribeWith()
                ?.topicFilter(topic)
                ?.send()
                ?.whenComplete { _, throwable ->
                    if (throwable != null) {
                        promise.reject("UNSUBSCRIBE_FAILED", throwable.message, throwable)
                    } else {
                        promise.resolve("Unsubscribed")
                        sendEvent("onMqttUnsubscribed", mapOf("topic" to topic))
                    }
                }
        }

        AsyncFunction("publish") { topic: String, message: String, qos: Int, retained: Boolean, promise: Promise ->
            val mqttQos = MqttQos.fromCode(qos) ?: MqttQos.AT_MOST_ONCE
            
            client?.publishWith()
                ?.topic(topic)
                ?.payload(message.toByteArray())
                ?.qos(mqttQos)
                ?.retain(retained)
                ?.send()
                ?.whenComplete { _, throwable ->
                    if (throwable != null) {
                        promise.reject("PUBLISH_FAILED", throwable.message, throwable)
                    } else {
                        promise.resolve("Published")
                    }
                }
        }
    }

    private fun scheduleReconnect() {
        val autoReconnect = storedOptions["autoReconnect"] as? Boolean ?: true
        if (!autoReconnect) return

        val delayMs = (storedOptions["reconnectDelay"] as? Number)?.toLong() ?: 5000L

        reconnectRunnable?.let { handler.removeCallbacks(it) }
        reconnectRunnable = Runnable {
            client?.let {
                sendEvent("onMqttReconnecting", mapOf("status" to "reconnecting"))
                // Reconnect doesn't have a simple .reconnect() in Mqtt3AsyncClient Async API easily without re-building
                // or calling connectWith() again with same options!
                val cleanSession = storedOptions["cleanSession"] as? Boolean ?: false
                val keepAlive = (storedOptions["keepAlive"] as? Number)?.toInt() ?: 60
                
                val connectBuilder = it.connectWith()
                    .cleanSession(cleanSession)
                    .keepAlive(keepAlive)
                
                // Re-apply auth if stored in class variables (not stored currently in options dictionary for safety)
                // Wait, I should store username/password or access them if needed for secret reuse, 
                // but the builder preserves them unless reset?
                // Actually with builder.buildAsync(), the client maintains its credentials if they were set on building!
                // Wait, credentials in connecting can be set on connectWith too!
                
                connectBuilder.send().whenComplete { ack, throwable ->
                    if (throwable == null) {
                        sendEvent("onMqttConnected", mapOf("status" to "connected"))
                        resubscribeAll()
                    }
                }
            }
        }
        handler.postDelayed(reconnectRunnable!!, delayMs)
    }

    private fun resubscribeAll() {
        for ((topic, qos) in subscribedTopics) {
            val mqttQos = MqttQos.fromCode(qos) ?: MqttQos.AT_MOST_ONCE
            client?.subscribeWith()
                ?.topicFilter(topic)
                ?.qos(mqttQos)
                ?.send()
        }
    }
}

File: modules/expo-native-mqtt/ios/ExpoNativeMqttModule.swift
import ExpoModulesCore
import CocoaMQTT

public class ExpoNativeMqttModule: Module {
  private var mqtt: CocoaMQTT?
  private var storedOptions: [String: Any] = [:]
  private var subscribedTopics: [String: Int] = [:] // topic -> qos
  private var reconnectTimer: Timer?
  private var isDisconnectingManually = false

  public func definition() -> ModuleDefinition {
    Name("ExpoNativeMqtt")

    Events(
      "onMqttConnected",
      "onMqttDisconnected",
      "onMqttError",
      "onMqttMessageReceived",
      "onMqttSubscribed",
      "onMqttUnsubscribed",
      "onMqttReconnecting"
    )

    AsyncFunction("connect") { (brokerUrl: String, username: String?, password: String?, options: [String: Any], promise: Promise) in
      guard let url = URL(string: brokerUrl) else {
        promise.reject("INVALID_URL", "Invalid broker URL")
        return
      }

      let host = url.host ?? brokerUrl
      let port = UInt16(url.port ?? 1883)
      
      self.storedOptions = options
      let clientId = options["clientId"] as? String ?? "expo-mqtt-\(UUID().uuidString)"
      let cleanSession = options["cleanSession"] as? Bool ?? false
      let keepAlive = options["keepAlive"] as? UInt16 ?? 60

      self.mqtt = CocoaMQTT(clientID: clientId, host: host, port: port)
      self.mqtt?.username = username
      self.mqtt?.password = password
      self.mqtt?.keepAlive = keepAlive
      self.mqtt?.cleanSession = cleanSession
      self.mqtt?.delegate = self

      // SSL Setup
      if brokerUrl.hasPrefix("ssl://") || brokerUrl.hasPrefix("wss://") {
        self.mqtt?.enableSSL = true
        self.mqtt?.allowUntrustCACert = true // match standard RN-Mqtt-Client behavior
      }

      self.isDisconnectingManually = false
      let success = self.mqtt?.connect() ?? false
      
      if success {
        promise.resolve("Connecting")
      } else {
        promise.reject("CONNECT_FAILED", "Failed into initiate connection")
      }
    }

    AsyncFunction("disconnect") { (promise: Promise) in
      self.isDisconnectingManually = true
      self.reconnectTimer?.invalidate()
      self.reconnectTimer = nil
      self.mqtt?.disconnect()
      promise.resolve("Disconnected")
    }

    AsyncFunction("subscribe") { (topic: String, qos: Int, promise: Promise) in
      self.subscribedTopics[topic] = qos
      let cocoaQos = CocoaMQTTQoS(rawValue: UInt8(qos)) ?? .qos0
      self.mqtt?.subscribe(topic, qos: cocoaQos)
      promise.resolve("Subscribed")
    }

    AsyncFunction("unsubscribe") { (topic: String, promise: Promise) in
      self.subscribedTopics.removeValue(forKey: topic)
      self.mqtt?.unsubscribe(topic)
      promise.resolve("Unsubscribed")
    }

    AsyncFunction("publish") { (topic: String, message: String, qos: Int, retained: Bool, promise: Promise) in
      let cocoaQos = CocoaMQTTQoS(rawValue: UInt8(qos)) ?? .qos1
      self.mqtt?.publish(topic, withString: message, qos: cocoaQos, retained: retained)
      promise.resolve("Published")
    }
  }

  fileprivate func scheduleReconnect() {
    guard !isDisconnectingManually else { return }
    let autoReconnect = storedOptions["autoReconnect"] as? Bool ?? true
    guard autoReconnect else { return }

    let delayMs = storedOptions["reconnectDelay"] as? Double ?? 5000.0
    let delaySeconds = delayMs / 1000.0

    DispatchQueue.main.async {
      self.reconnectTimer?.invalidate()
      self.reconnectTimer = Timer.scheduledTimer(withTimeInterval: delaySeconds, repeats: false) { [weak self] _ in
        guard let self = self else { return }
        self.sendEvent("onMqttReconnecting", ["status": "reconnecting"])
        _ = self.mqtt?.connect()
      }
    }
  }

  fileprivate func resubscribeAll() {
    for (topic, qos) in subscribedTopics {
      let cocoaQos = CocoaMQTTQoS(rawValue: UInt8(qos)) ?? .qos0
      mqtt?.subscribe(topic, qos: cocoaQos)
    }
  }
}

extension ExpoNativeMqttModule: CocoaMQTTDelegate {
  public func mqtt(_ mqtt: CocoaMQTT, didConnectAck ack: CocoaMQTTConnAck) {
    if ack == .accept {
      sendEvent("onMqttConnected", ["status": "connected"])
      resubscribeAll()
    } else {
      sendEvent("onMqttError", ["error": "Connection refused: \(ack)"])
    }
  }

  public func mqtt(_ mqtt: CocoaMQTT, didDisconnectWithError error: Error?) {
    sendEvent("onMqttDisconnected", [
      "error": error?.localizedDescription ?? "clean disconnect"
    ])
    scheduleReconnect()
  }

  public func mqtt(_ mqtt: CocoaMQTT, didReceiveMessage message: CocoaMQTTMessage, id: UInt16) {
    sendEvent("onMqttMessageReceived", [
      "topic": message.topic,
      "payload": message.string ?? "",
      "qos": Int(message.qos.rawValue),
      "retained": message.retained
    ])
  }

  public func mqtt(_ mqtt: CocoaMQTT, didSubscribeToTopic topics: [String], id: UInt16) {
    sendEvent("onMqttSubscribed", ["topics": topics])
  }

  public func mqtt(_ mqtt: CocoaMQTT, didUnsubscribeFromTopic topic: String) {
    sendEvent("onMqttUnsubscribed", ["topic": topic])
  }

  public func mqtt(_ mqtt: CocoaMQTT, didPublishMessage message: CocoaMQTTMessage, id: UInt16) {
    // Optional: could emit event, but AsyncFunction resolves immediately
  }

  public func mqtt(_ mqtt: CocoaMQTT, didPublishAck id: UInt16) { }

  public func mqtt(_ mqtt: CocoaMQTT, didSubscribeToTopics topics: [String], id: UInt16) { }

  public func mqttOneMinute(_ mqtt: CocoaMQTT) { }
}
